generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,157 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by bitwise shifting left or right by an
+   * arbitrary number of bits. Generates an opaque value if the type of zero to be generated is an
+   * integer and the shading language in use supports bitwise operations.
+   *
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // We need to make sure the LHS and RHS of the shift expression are based on the same type
+    // explicitly, so we can't generate opaque values to shift by as they will implicitly be
+    // considered signed integers, and both sides of the shift must be the same type.
+    final Expr shiftValue;
+    if (isZero) {
+      // We can't shift zero reliably if the number is signed because of sign bit extension, so
+      // we simply shift it zero if it's signed.
+      shiftValue = type.getElementType() == BasicType.INT
+          ? new IntConstantExpr("0")
```

Comment:
I don't think it is safe to use a random value in the range [0, 1000).  The spec says:

"The result is undefined if the right operand is negative, or greater than or
equal to the number of bits in the left expression’s base type."

Now, what can we assume about the number of bits in the left expression's base type?  Without some precision analysis, we should conservatively assume that it might be a lowp integer.  The spec says:

"lowp integers (signed and unsigned) must be represented as an integer with between 9 and 32 bits
inclusive."

So to be safe we should make the RHS be in the range [0, 9).

Reply:
Done.